src/plugins/intel_cpu/src/nodes/memory_state_base.cpp

```diff
@@ -14,6 +15,8 @@ MemoryNode::MemoryNode(const std::shared_ptr<ov::Node>& op) {
         m_id = assignOp->get_variable_id();
     } else if (auto readValueOp = ov::as_type_ptr<ov::op::util::ReadValueBase>(op)) {
         m_id = readValueOp->get_variable_id();
+    } else if (auto readValueWithSubgraphOp = ov::as_type_ptr<ov::intel_cpu::ReadValueWithSubgraphNode>(op)) {
+        m_id = readValueWithSubgraphOp->get_variable_id();
```

Comment:
The fact that `util::VariableExtension::get_variable_id()` is a pure method is not an issue at all, this is how the interfaces work in cpp. All you need is instead of casting to the dynamic type cast the pointer to the interface type and use the interface method.
Here is an example:
https://github.com/openvinotoolkit/openvino/blob/7eef06909d0c9a5932ad97f810a2d1c5281d164b/src/inference/src/dev/iplugin.cpp#L218-L231

Reply:
Updated, Just also removed
```
    if (auto assignOp = ov::as_type_ptr<ov::op::util::AssignBase>(op)) {
        m_id = assignOp->get_variable_id();
    } else if (auto readValueOp = ov::as_type_ptr<ov::op::util::ReadValueBase>(op)) {
        m_id = readValueOp->get_variable_id();
```
only keep one:
`if (auto assignOp = std::dynamic_pointer_cast<ov::op::util::VariableExtension>(op)) {`